### PR TITLE
xkb: implementation of keyboard layout switcher

### PR DESCRIPTION
**Pull request by lexa**
This patch adds basic support for X KeyBoard extensions. It allows to get/set current keyboard layout.

This patch adds widget to display current keyboard layout, it displays
layout short name. Current layout can be changed either by clicking on
widget or by hotkey.

Current implementation doesn't support any configuration of layout,
it's a merely a layout indicator and switcher, however layout
configuration can be set by tools like setxkbmap or by any third-party
tools.

Indicators looks like a short name of layout (two letters usually), and looks like this:
layout_indicator_example

### CODE REVIEW

**Comment by Elv13** [comment]
Hello,

First thing first. Could you split the PR in 2? One for the backend part and (later) one for the widget part?

I don't think adding that many line to rc.lua is a good idea. Many people (lucky them) also don't change keyboard layout at all. Maybe is widget in awful.widgets would be the proper place for the ui code.

Thanks

**Comment by blueyed** [comment]
> I don't think adding that many line to rc.lua is a good idea.

Agreed.  The changes to the default config should be as few as possible.

**Comment by lexa** [comment]
Thank you for your comments.
>The changes to the default config should be as few as possible.
I agree with you, could you please look at updated merge request. 
I have split commit in two:
in first commit only changes in backend. Second commit adds widget for displaying keyboard layout.
Also changes in default config minimised.

**Comment by Elv13** [comment]
Already much better!

Now the question is, given that by default it will alwyas have a single layout, is it worth adding to the default rc.lua? The pros are that it help discoverability. The cons is that, by itself, until you configure it, it is useless (beside knowing the current layout). I personally don't mind much one way or another.

@blueyed  @psychon: What do you think?

**Comment by lexa** [comment]
>Now the question is, given that by default it will alwyas have a single layout, is it worth adding to the >default rc.lua?

I think that widget can be hidden if only one layout configured, and when user configure his layout then widget can be shown. What do you think about that idea?

**Comment by blueyed** [comment]
Yes, I prefer hiding it by default when there is only a single layout.

**Comment by lexa** [comment]
Please, look at updated patch, i made changes as you requested. In case of single layout indicator won't show up.

File: awesomerc.lua.in [hunk 1]
```diff
@@ -270,7 +276,9 @@ globalkeys = awful.util.table.join(
                   awful.util.getdir("cache") .. "/history_eval")
               end),
     -- Menubar
-    awful.key({ modkey }, "p", function() menubar.show() end)
+    awful.key({ modkey }, "p", function() menubar.show() end),
+    -- Alt + Left Shift switches the current keyboard layout
+    awful.key({ "Mod1" }, "Shift_L", function () mykeyboardlayout.next_layout() end)
```
Comment (Elv13) [inline awesomerc.lua.in]: I disagree with `Alt+Shift` as those are 2 modifiers, it will cause problems with everybody who use Alt as their modkey (mostly former DWM users), Mac OSX (and my config) use Alt+Space. I understand Shift is used by Windows and is quite convenient to use, but it may cause a little backsplash.

Other than that, this patch now look fine to me. I will test it later today then, unless someone else complain, merge it.

Comment (psychon) [inline awesomerc.lua.in]: Ok, will take a quick look. That should be enough to delay things until I have some more time. Sorry!

Comment (lexa) [inline awesomerc.lua.in]: I see your point. I will not change default hotkeys. Let user choose appropriate hotkey.

**Comment by lexa** [comment]
I'm appreciate your effort on commenting this patch. I hope i correctly fixed all issues on which you pointed me. Could you please take a look on updated merge request?

**Comment by lexa** [comment]
I'm sorry to bother you, do you have any comments on this pull request preventing it from been merged ?

**Comment by blueyed** [comment]
@lexa
Sorry for letting you wait so long.  I had skimmed it earlier today, but a) had not enough time to look really into it, and b) lack the experience with the stuff you're touching.

I think that in the end @psychon should review it.

**Comment by psychon** [comment]
I knew it would end up being my job. :-(

The C part looks good. I still don't know much about the XKB extension, but I trust @lexa here and this isn't "that much API".

The Lua parts you can review, too. :-P My only comment from a quick look would be to handle the `connect_signal` part differently. This effectively prevents any of these widgets to ever be garbage collected and it isn't really necessary since all these widgets are the same. Perhaps this can be replaced with a singleton which is lazily initialized the first time the widget is used...? Or am I over-engineering things?

TL;DR: The C part looks good enough to me, others decide about the lua part (sorry!).

**Comment by lexa** [comment]
>My only comment from a quick look would be to handle the connect_signal part differently. This effectively prevents any of these widgets to ever be garbage collected and it isn't really necessary since all these widgets are the same. Perhaps this can be replaced with a singleton which is lazily initialized the first time the widget is used...?

I hope that grasped your idea correctly, widget 'keyboardlayout' will be created only once and all following attempts of create it will return previously created instance.

**Comment by psychon** [comment]
What do others think? @blueyed @Elv13

**Comment by Elv13** [comment]
Look good to me

**Comment by Elv13** [comment]
There is no point in waiting for longer, issues will be fixed if/when they arise.

Thanks for this!

**Comment by actionless** [comment]
when building from AUR (https://aur.archlinux.org/packages/awesome-git/) ldoc task is failing:
```
Scanning dependencies of target ldoc
reading configuration from config.ld
module(...) name deduction failed: base /tmp/yaourt-tmp-lie/aur-awesome-git/src/build/docs/capi/ /tmp/yaourt-t
mp-lie/aur-awesome-git/src/build/lib/awful/widget/keyboardlayout.lua
CMakeFiles/ldoc.dir/build.make:141: recipe for target 'CMakeFiles/ldoc' failed
make[2]: *** [CMakeFiles/ldoc] Error 255
CMakeFiles/Makefile2:224: recipe for target 'CMakeFiles/ldoc.dir/all' failed
make[1]: *** [CMakeFiles/ldoc.dir/all] Error 2
Makefile:116: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by blueyed** [comment]
The requirement `xcb-xkb>=1.11` cannot be fulfilled on Ubuntu/Debian (which has 1.10 only).

1. Is 1.11 really necessary?
2. Can this be made optionally, which may be useful if you want to skip the xcb-xkb dependency altogether?

**Comment by Elv13** [comment]
I expected that kind of issues, but at this point, as the feature is quite useful, I didn't want to let this rot any longer. We had the same issues with `xcb-cursor` last year and we did not made it optional then. Yes, it will cause issues, but I guess it is necessary.

**Comment by blueyed** [comment]
I see your point, but would not like awesome being difficult to build because of a keyboard layout switcher - especially if xcb-xkb>=1.11 wasn't necessary.
The changelog does not look like it's a breaking release: http://lists.x.org/archives/xorg-announce/2014-August/002469.html.

**Comment by blueyed** [comment]
@actionless 
This is fixed in 98c4ae8 now.

**Comment by blueyed** [comment]
I've removed the version dependency for xcb.xkb in https://github.com/awesomeWM/awesome/commit/d2ddb3370830241c8ec3b5a23e3055a6b7f9f299.
While I still haven't tested the layout switcher itself, it now is easier to build for the moment again.
